h2o-algos/src/main/java/hex/util/LinearAlgebraUtils.java

```diff
@@ -164,23 +166,38 @@ public BMulInPlaceTask(DataInfo xinfo, double[][] yt, int nColsExp) {
 
     public double[][] _atq;    // Output: A'Q is p_exp by k, where p_exp = number of cols in A with categoricals expanded
 
-    public SMulTask(DataInfo ainfo, int ncolQ) {
+    public SMulTask(DataInfo ainfo, int ncolQ, int ncolExp) {
       _ainfo = ainfo;
       _ncolA = ainfo._adaptedFrame.numCols();
-      _ncolExp = numColsExp(ainfo._adaptedFrame,true);
+      _ncolExp = ncolExp;   // when call from GLRM or PCA
       _ncolQ = ncolQ;
+      _atq = new double[_ncolExp][_ncolQ];
```

Comment:
This looks wrong, unless intentional. Allocating array in constructor will mean the array is *shared* among the threads. Any modification is then racy. If this was not intentional it would also suggest missing multi chunk correctness test.